src/Npgsql/NpgsqlWriteBuffer.cs

```diff
@@ -280,11 +274,17 @@ public void WriteDouble(double value, bool littleEndian)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         void Write<T>(T value)
         {
-            Debug.Assert(Unsafe.SizeOf<T>() <= WriteSpaceLeft);
+            if (Unsafe.SizeOf<T>() > WriteSpaceLeft)
+                ThrowNotSpaceLeft();
+
             Unsafe.WriteUnaligned(ref Buffer[WritePosition], value);
             WritePosition += Unsafe.SizeOf<T>();
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Out of curiosity, why NoInlining? Is it to keep the calling methods small, so that they themselves can be inlined? Would it make sense to instead have a method that returns an exception and have a throw statement in the calling methods?

Reply:
I will remove it since the JIT is smart enough to not inline it. The reason to move exception throwing logic away is to have short hot paths. This allows to the minimize assembly size and to produce smaller JIT compiled methods.

You can read about this magic trick in the CoreFX repo in comments of some throw helpers. For example, from [here](https://github.com/dotnet/corefx/blob/56dfb8834fa50f3bc61ea9b4bfdc9dcc759b6ec9/src/System.Memory/src/System/ThrowHelper.cs#L11-L22):

> This pattern of easily inlinable "void Throw" routines that stack on top of NoInlining factory methods is a compromise between older JITs and newer JITs (RyuJIT in Core CLR 1.1.0+ and desktop CLR in 4.6.3+).

> This package is explicitly targeted at older JITs as newer runtimes expect to implement Span intrinsically for best performance.
> The aim of this pattern is three-fold
> 1. Extracting the throw makes the method preforming the throw in a conditional branch smaller and more inlinable
> 2. Extracting the throw from generic method to non-generic method reduces the repeated codegen size for value types
> 3a. Newer JITs will not inline the methods that only throw and also recognise them, move the call to cold section and not add stack prep and unwind before calling https://github.com/dotnet/coreclr/pull/6103
> 3b. Older JITs will inline the throw itself and move to cold section; but not inline the non-inlinable exception factory methods - still maintaining advantages 1 & 2